sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -834,8 +874,29 @@ object DateTimeUtils {
     }
   }
 
+  private def truncTime(ts: SQLTimestamp, level: Int): SQLTimestamp = {
+    val unitInMillis = level match {
+      case TRUNC_TO_DAY => MILLIS_PER_DAY
+      case TRUNC_TO_HOUR => MILLIS_PER_HOUR
+      case TRUNC_TO_MINUTE => MILLIS_PER_MINUTE
+      case TRUNC_TO_SECOND => MILLIS_PER_SECOND
+      case _ =>
+        // caller make sure that this should never be reached
+        sys.error(s"Invalid trunc level: $level")
+    }
+
+    val millisUtc = ts / 1000L
+    val millisLocal = millisUtc + threadLocalLocalTimeZone.get().getOffset(millisUtc)
+    val days = Math.floor(millisLocal.toDouble / unitInMillis).toInt
```

Comment:
This is not days right?

Are we using the Math.floor to prevent rounding problems if millisLocal < 0?